include/seqan3/range/decorator/gap_decorator.hpp

```diff
@@ -576,12 +178,16 @@ ame Arithmetic operators
      * Average and worst case (insertion before last gap): \f$O(k)\f$,
      * Best case (back insertion): \f$O(\log k)\f$.
      */
-    iterator insert_gap(iterator const it, size_type const count = 1)
+    template <typename iterator_t>
+    iterator_t insert_gap(iterator_t const it, size_type const count = 1)
```

Comment:
I think we should be as explicit in public api as possible. This signature suggest that we accept ANY iterator.

I asked this further down, but why did we inherit the bidrectional from the random_access one again? I forgot which benefit it had.

I personally would prefer a signature like this. Because this will also generate nice documentation.

```c++
template <typename iterator_category>
gap_decorator_iterator<iterator_category> insert_gap(gap_decorator_iterator<iterator_category> const it, size_type const count = 1)
```

Reply:
> I asked this further down, but why did we inherit the bidrectional from the random_access one again? I forgot which benefit it had.

Because we want the normal iterator do all the things a random access iterator can without being a random access iterator. So that the default is that we get the slow performance in generic algorithms because people might expect something faster if they require random access but they won't get it with constant time access. This is not my opinion especially since the decorator is trying to emulate a sequence and jumping in it seems just natural. But it was argued that the std::set also does not provide random access with logarithmic performance.